src/_pytest/assertion/util.py

```diff
@@ -373,7 +377,10 @@ def _compare_eq_sequence(
                 left_value = left[i]
                 right_value = right[i]
 
-            explanation += [f"At index {i} diff: {left_value!r} != {right_value!r}"]
+            explanation.append(
+                f"At index {i} diff:"
+                f" {highlighter(repr(left_value)).strip()} != {highlighter(repr(right_value)).strip()}"
```

Comment:
The `highlight()` from pygments always finished by a newline.

Alternatively, I can make change the hightlighter to be 

```
def highlight(...):
    return writer._highlight(...).strip()
```

so we only do it in one place

Reply:
I think it's worth fixing in TerminalWriter itself, e.g. a diff like this (untested):

```diff
diff --git a/src/_pytest/_io/terminalwriter.py b/src/_pytest/_io/terminalwriter.py
index 934278b93..4ac578b6d 100644
--- a/src/_pytest/_io/terminalwriter.py
+++ b/src/_pytest/_io/terminalwriter.py
@@ -223,6 +223,10 @@ class TerminalWriter:
                         style=os.getenv("PYTEST_THEME"),
                     ),
                 )
+                # Pygments seems to add a newline even when not present in the
+                # source, we need it to be the same.
+                if highlighted.endswith("\n") and not source.endswith("\n"):
+                    highlighted = highlighted[:-1]
                 return highlighted
             except pygments.util.ClassNotFound:
                 raise UsageError(
```

In any case we shouldn't `strip()` as it might remove *too* many newlines.

Reply:
That's fair.

I went for this implementation though it is not exactly perfect, as there could be multiple newlines and we might end up with n+1, though I have not seen this be the case, so should be fine